braindecode/datautil/windowers.py

```diff
@@ -0,0 +1,200 @@
+"""Get epochs from mne.Raw
+"""
+
+# Authors: Hubert Banville <hubert.jbanville@gmail.com>
+#          Lukas Gemein <l.gemein@gmail.com>
+#          Simon Brandt <simonbrandt@protonmail.com>
+#          David Sabbagh <dav.sabbagh@gmail.com>
+#
+# License: BSD (3-clause)
+
+import numpy as np
+import mne
+import pandas as pd
+
+
+class Windower(object):
+    """
+    A windower that creates a mne Epochs objects.
+    """
+    def __init__(self, trial_start_offset_samples, trial_stop_offset_samples,
+                 supercrop_size_samples, supercrop_stride_samples,
+                 drop_samples=False, mapping=None):
+        self.trial_start_offset_samples = trial_start_offset_samples
+        self.trial_stop_offset_samples = trial_stop_offset_samples
+        assert supercrop_size_samples > 0, (
+            "supercrop size has to be larger than 0")
+        self.size = supercrop_size_samples
+        assert supercrop_stride_samples > 0, (
+            "supercrop stride has to be larger than 0")
+        self.stride = supercrop_stride_samples
+        self.drop_samples = drop_samples
+        self.mapping = mapping
+        # TODO: assert values are integers
+        # TODO: assert start < stop
+
+    # TODO: handle case we don't get a raw
+    def __call__(self, raw, events, metadata):
+        # supercrop size - 1, since tmax is inclusive
+        return mne.Epochs(raw, events, baseline=None,
+                          tmin=0, tmax=(self.size-1)/raw.info["sfreq"],
+                          metadata=metadata)
+
+
+class EventWindower(Windower):
+    """
+    A Windower that creates supercrops/windows based on events in mne.Raw.
+    Therefore, it fits supercrops of supercrop_size_samples in
+    trial_start_offset_samples to trial_stop_offset_samples separated by
+    supercrop_stride_samples. If the last supercrop does not end
+    at trial_stop_offset_samples, creates another overlapping supercrop that
+    ends at trial_stop_offset_samples if drop_samples is set to False.
+
+    in mne: tmin (s)                    trial onset        onset + duration (s)
+    trial:  |--------------------------------|--------------------------------|
+    here:   trial_start_offset_samples                trial_stop_offset_samples
+
+    Parameters
+    ----------
+    trial_start_offset_samples: int
+        start offset from original trial onsets in samples
+    trial_stop_offset_samples: int
+        stop offset from original trial onsets in samples
+    supercrop_size_samples: int
+        supercrop size
+    supercrop_stride_samples: int
+        stride between supercrops
+    drop_samples: bool
+        whether or not have a last overlapping supercrop/window, when
+        supercrops/windows do not equally devide the continuous signal
+    mapping: dict(str: int)
+        mapping from event description to target value
+    """
+    def __call__(self, base_ds):
```

Comment:
Why do we pass a Dataset object here instead of a mne.Epochs object? Wouldn't it be simpler to directly work with the Raw object?

Reply:
In the case of the FixedLenthWidnower, we need additional information that is not contained in the Raw object. Since there are no events defined for the TUH data, the target is provided in the BaseDataset on a dataset level. We have to retrieve this information when cutting supercrops/windows, to assign a target to every supercrop.

Reply:
Can you give examples of targets that you would use on the TUH data (I guess normal/abnormal, age,sex, etc.)? I was thinking there might be a way to save that data inside the Raw.info structure - in that case, we could work directly with a Raw object an avoid another level of abstraction. The FixedLengthWindower could extract all the fields we care about from the info structure and put them in the `metadata` attribute of the Epochs object.

@agramfort what do you think about adding other fields to the Raw.info structure to pass other recording-level labels (e.g., normal/abnormal)?

Reply:
The dataset was designed to decode normal/abnormal. However, we have also used age and sex of patients in some experiments. I guess, these are the obvious decoding choices. You are right, adding information to fields in mne Raw objects is another way of doing this. However, as I said above, it might be an advanage to have this info in a DataFrame for convenient data splits. For example, maybe you want to decode the pathology status only from recordings of patients that exceed a certain age.